[PR-1251] Article about the state of doctests in Haskell (#9)
serokell/blog-posts: introduction-to-doctests-in-haskell/article.md

# Introduction to Doctests in Haskell

If you ever had to skim through the Haskell source code documentation, you might have noticed those lines with fancy `>>>` symbols at the beginning.
They are called 'doctests' or 'doctest examples'.

To put it simply, doctests are pieces of text within [Haddock comments](https://www.haskell.org/haddock/) that look like interactive Haskell sessions.
The idea to search for and execute those sessions all over Haskell modules actually comes from the awesome Python's
[`doctest`](https://en.wikipedia.org/wiki/Doctest) module.

Documenting software is extremely challenging.
Doctests make the documentation process not only pleasant and effective – it actually allows Haskell functions to speak for themselves – but
also helps to get the most out of your testing efforts.

## How to define a doctest?

> Let's look at an example of a doctest. 

Reviewer: Are there any other doctest libraries? Do they all follow the same syntax?

Author: There're a few more libraries. But, AFAIK, they do follow the same syntax

> ```haskell

Reviewer: Want to perhaps take something from Haskell source? 

Like:
```haskell
-- | @const x@ is a unary function which evaluates to @x@ for all inputs.
--
-- >>> const 42 "hello"
-- 42
--
-- >>> map (const 42) [0..3]
-- [42,42,42,42]
```
Here, the lack of comment and the <BLANKLINE> seems a bit confusing to me, personally.

Reviewer: I've added this particular piece a bit higher, let me know if you don't like that.

Author: I do agree with the changes :)

-- |
-- >>> 1 + 2
-- 3
```

```haskell
-- |
-- >>> putStrLn ""
-- <BLANKLINE>
```

These are some basic markup requirements:

> * Every doctest example should be placed within a valid piece of Haddock documentation, that is followed by either special `-- |` or `{- |` syntaxes.

Reviewer: Is the `-- |` or `{-|` just for signifying Haddock documentation?

Reviewer: Checked Haddock docs myself. ;D

# Introduction to Doctests in Haskell

Documenting software can be challenging, but it doesn't always need to be so.

> In this article, we'll introduce a concept – doctests – that makes the documentation process pleasant and effective.

Reviewer 2: Nitpick: I think this text flows a bit better in my mind:

In this article, we'll introduce doctests: a concept that makes the documentation process pleasant and effective.

Reviewer: Agree.

cabal-docspec
```

The command above should fail with the following error:

```
expected: "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
but got:
            ^
            <interactive>:10:1: error:
                Variable not in scope: encode :: Anime -> t
```

The above happens because the library handles things a little differently – it requires modules to be explicitly imported/exported.

Let's edit our doctests a little to make `cabal-docspec` work:

```diff
 -- |
+ -- >>> import Data.Aeson
 -- >>> encode favourite
 -- "{\"title\":\"One-Punch Man\",\"rating\":8.9}"
 --
 favourite :: Anime
 favourite = Anime "One-Punch Man" 8.9

```

Now the `cabal-docspec` command should succeed:

```none
Total:         2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
Examples:      2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
```

#### Why you should use `cabal-docspec`

Here are the reasons why you might want to use `cabal-docspec` for your project: 

* The library is a lot faster than the alternatives because it uses compiled code.

* The library doesn't depend on GHC as a library, so it's much more resilient to GHC version changes.

> * The library doesn't require recompiling the source code if it's only about writing doctests – this saves additional time for developers and

Reviewer: ```suggestion
* The library doesn't require recompiling the source code if you only change the doctests – this saves additional time for developers and
```

Is this what you meant?

Author: Yup, exactly
